### PR TITLE
New Shibboleth authentication

### PR DESCRIPTION
**Pull request by raphendyr**
* Dictionary style settings for shibboleth configuration
* Implement a parser for shib2 encoded multivalue fields.
* Parse only a single student number, based on a config value
* Authentication backend is rewritten, but should be more robust
  and handle problems better than before.
  e.g. before, oo long input values were just cut, and thus multiple
  different inputs might have resulted same output.
  Now we raise errors instead and require someone to fix the actual
  problem.
* Better debug output.. So we can actually find out these problems :)

Note that the `SHIBBOLETH_ENVIRONMENT_VARS.STUDENT_DOMAIN` configuration variable is now required with shibboleth! Check the documentation.

### CODE REVIEW

**Comment by raphendyr** [comment]
retest this please

**Comment by raphendyr** [comment]
retest this please

**Comment by raphendyr** [comment]
retest this please

File: shibboleth_login/auth_backend.py [hunk 1]
```diff
@@ -35,62 +23,138 @@ def authenticate(self, request, shibd_meta=None):
...
+            first_, __, last_ = full_name.partition(' ')
+            if not user.first_name:
+                user.first_name = first_[:first_name_len]
+            if not user.last_name:
+                user.last_name = last_[:last_name_len]
+            user_save_flag = True
+
         if user_save_flag:
+            # TODO: write better error reporting, when there is a validator to raise something
+            user.clean()
```
Comment (Mankro) [inline shibboleth_login/auth_backend.py]: Shouldn't the `full_clean()` method be used instead of `clean()`?

File: shibboleth_login/parser.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+                    "Content is %s" % (name, fields))
+            ns, _, value = field[3:].partition(urn)
+            values = tuple(reversed(value.split(':'))) + (ns.lstrip(':'), field)
+            if filters and not all(values[i] == v for i, v in filters.items()):
+                continue
+            data.append(values)
+        return data
+
+
+class HakaParser(Parser):
```
Comment (Mankro) [inline shibboleth_login/parser.py]: Is `HakaParser` actually used anywhere?

File: shibboleth_login/auth_backend.py [hunk 3]
```diff
@@ -35,62 +23,138 @@ def authenticate(self, request, shibd_meta=None):
...
 
+        # TODO: support multiple domains
         profile = user.userprofile
-        student_id = self._parse(shibd_meta, STUDENT_ID_KEY)
-        if student_id:
-            student_id = student_id.split(':')[-1][:25]
-            if student_id and student_id != profile.student_id:
-                profile.student_id = student_id
-                profile.save()
+        sid_filters = env_settings.STUDENT_FILTERS.copy()
```
Comment (Mankro) [inline shibboleth_login/auth_backend.py]: Just to check, does the parser for student ids work with non-integer values? Some student ids mix digits and letters. The long Shibboleth URN seems to include the word `int` even when the value really isn't an integer.

**Comment by jykesaviaGNH** [comment]
I can confirm that there's data around (e.g. "A57575757") with still the claim 'int' in the field.


On 31.7.2019 12.44, Markku Riekkinen wrote:

@Mankro commented on this pull request.

________________________________

In shibboleth_login/auth_backend.py<https://github.com/Aalto-LeTech/a-plus/pull/392#discussion_r309130075>:

>          profile = user.userprofile
-        student_id = self._parse(shibd_meta, STUDENT_ID_KEY)
-        if student_id:
-            student_id = student_id.split(':')[-1][:25]
-            if student_id and student_id != profile.student_id:
-                profile.student_id = student_id
-                profile.save()
+        sid_filters = env_settings.STUDENT_FILTERS.copy()


Just to check, does the parser for student ids work with non-integer values? Some student ids mix digits and letters. The long Shibboleth URN seems to include the word int even when the value really isn't an integer.

File: shibboleth_login/auth_backend.py (hunk 3 above)
Comment (raphendyr) [reply shibboleth_login/auth_backend.py]: I think the `int` refers to international in this case. I thought about integers first too.

File: shibboleth_login/auth_backend.py (hunk 1 above)
Comment (raphendyr) [reply shibboleth_login/auth_backend.py]: Yes. Fix coming up.

File: shibboleth_login/parser.py (hunk 2 above)
Comment (raphendyr) [reply shibboleth_login/parser.py]: Nope. Was used in tests under `if __name__ == '__main__'`, but that was removed later on. Thanks for noticing this.

**Comment by raphendyr** [comment]
Updated. If no comments, I'll merge it.

**Comment by Mankro** [comment]
I think it's ready.

**Comment by raphendyr** [comment]
Just reminding that the `SHIBBOLETH_ENVIRONMENT_VARS.STUDENT_DOMAIN` configuration variable is now required with shibboleth! Check the documentation.

**Pull request merged**